### PR TITLE
📖 : Write tests for your own plugin

### PR DESCRIPTION
**Pull request by Kavinjsir**
### Description

Add `Write tests for your own plugin` section to introduce:
- [x] How to write e2e tests ([Preview](https://deploy-preview-2934--kubebuilder.netlify.app/plugins/creating-plugins.html#test-plugins))
- [x] How to generate samples under `./testdata/ `
- [x] Describe Kb plugin testing convention([CONTRIBUTING.md](https://github.com/Kavinjsir/kubebuilder/blob/docs%2Ftest-plugin/CONTRIBUTING.md#test-plugin))

### Motivation
Resolve https://github.com/kubernetes-sigs/kubebuilder/issues/2677

### Preivew
- [Guidance of writing tests](https://deploy-preview-2934--kubebuilder.netlify.app/plugins/testing-plugins.html)
- [Update create-plugin](https://deploy-preview-2934--kubebuilder.netlify.app/plugins/creating-plugins.html)

### CODE REVIEW

File: docs/book/src/SUMMARY.md [hunk 1]
```diff
@@ -118,6 +118,7 @@
   - [Extending the CLI](./plugins/extending-cli.md)
   - [Creating your own plugins](./plugins/creating-plugins.md)
   - [Plugins Versioning](./plugins/plugins-versioning.md)
+  - [Write Tests For Your Own Plugin](./plugins/test-plugin.md)
```
Comment (camilamacedo86) [inline docs/book/src/SUMMARY.md:121]: What about it be under `Creating your own plugins`

File: docs/book/src/plugins/test-plugin.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Write Tests For Your Own Plugins
+
+The Kubebuilder project tests plugins in multiple perspectives.
+
+First, the plugin's functionality is verified through E2E tests which will be performed in clusters of different K8s versions.
+
+Secondly, the scaffolded output from the plugin should be in consistent with multiple platforms.
```
Comment (camilamacedo86) [inline docs/book/src/plugins/test-plugin.md:7]: Hi @Kavinjsir see that KB can be used as a LIB.
Therefore, is possible to import KB in another repo and create your own plugin.
Then, after create your own plugin you will want to test that.
So, that we can share with the authors / its consumers that KB lib also provide the features to help out and then can do the tests such as / similarly we do in KB repo but in their own thing.

That is the goal of this doc. Is that make sense?

Comment (camilamacedo86) [reply docs/book/src/plugins/test-plugin.md:7]: However, we also could have a section like CONTRIBUTION where we could add the info on how to create new plugins on KB repo, how to test them out and etc.. that would be amazing actually.

Comment (Kavinjsir) [reply docs/book/src/plugins/test-plugin.md:7]: Yeah, that makes sense. Thx for clarifying!

File: docs/book/src/plugins/creating-plugins.md [hunk 3]
```diff
@@ -25,54 +25,93 @@ If you do not want to develop your plugin using Golang, you can follow its stand
...
-[new-plugin-gen]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/code-generate-image-plugin.md
-[kubebuilder-declarative-pattern]: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern
-[declarative-code]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/declarative
-[sdk]: https://github.com/operator-framework/operator-sdk
\ No newline at end of file
+- [New Plugin (`deploy-image.go.kubebuilder.io/v1beta1`) to generate code][new-plugin-gen]: its purpose is to provide an arch-type that will scaffold the APIs and Controllers with the required code to deploy and manage solutions on the cluster.
+
+Please, feel to contribute with them as well. Your contribution to the project is very welcome.
+
+- [sdk-cli-pkg](https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go)
```
Comment (Kavinjsir) [inline docs/book/src/plugins/creating-plugins.md:104]: Seems like the following links are missing in the web page: https://book.kubebuilder.io/plugins/creating-plugins.html#future-vision-for-kubebuilder-plugins

Do we want to display them? Just convert them to be visible as a list:
https://deploy-preview-2934--kubebuilder.netlify.app/plugins/creating-plugins.html#future-vision-for-kubebuilder-plugins

How would you like this?

File: docs/book/src/plugins/creating-plugins.md [hunk 4]
```diff
@@ -25,54 +25,93 @@ If you do not want to develop your plugin using Golang, you can follow its stand
...
+   - Call Kubectl to validate running resources. See [utils.Kubectl](https://pkg.go.dev/sigs.k8s.io/kubebuilder/v3/test/e2e/utils#Kubectl)
+4. Delete temporary resources after testing exited, e.g:
+   - Uninstall prerequisites CRDs: See [UninstallPrometheusOperManager](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.6.0/test/e2e/utils/test_context.go#L183)
+   - Delete temp dir. See [Destroy](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.6.0/test/e2e/utils/test_context.go#L255)
+
+**References:** [operator-sdk e2e tests](https://github.com/operator-framework/operator-sdk/tree/master/test/e2e/go), [kubebuiler e2e tests](https://github.com/kubernetes-sigs/kubebuilder/tree/master/test/e2e/v3)
+
+### Generate Test Samples
+
+TBD
```
Comment (Kavinjsir) [inline docs/book/src/plugins/creating-plugins.md:92]: @camilamacedo86  I might need some help here.

See that [Suggestion (Acceptence Criteria) ](https://github.com/kubernetes-sigs/kubebuilder/issues/2677#issue-1228818301) has mentioned generating samples in testdata.

If we're using kubebuilder as a lib, then are we going to use `TestContext` also to create sample projects? OR shall we encourage the operator authors to do similar scripting as [we did in kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.6.0/test/testdata/generate.sh)?

File: docs/book/src/SUMMARY.md (hunk 1 above)
Comment (Kavinjsir) [reply docs/book/src/SUMMARY.md]: Yeah, sounds good. 👍🏼  I misunderstood the origin idea previously...

File: docs/book/src/plugins/creating-plugins.md (hunk 4 above)
Comment (Kavinjsir) [reply docs/book/src/plugins/creating-plugins.md]: Currently left description of **using TextContext** to generate sample projects.

If using `$kb` command is more recommended,(similar as what Kubebuilder did in testdata) I can rewrite this section.

See: [Preview](https://deploy-preview-2934--kubebuilder.netlify.app/plugins/creating-plugins.html#generate-test-samples)

File: docs/book/src/plugins/test-plugin.md (hunk 2 above)
Comment (Kavinjsir) [reply docs/book/src/plugins/test-plugin.md]: Hi @camilamacedo86 . I tried adding a section under [CONTRIBUTING.md](https://github.com/Kavinjsir/kubebuilder/blob/docs%2Ftest-plugin/CONTRIBUTING.md#test-plugin) to describe the testing new plugins on KB repo. Not sure if it is suitable though...

File: docs/book/src/plugins/creating-plugins.md (hunk 3 above)
Comment (Kavinjsir) [reply docs/book/src/plugins/creating-plugins.md:150]: Sounds good! Yeah, I like the idea of a new KubeGen Project. I'll remove this entire section now.

File: docs/book/src/plugins/creating-plugins.md [hunk 5]
```diff
@@ -25,54 +25,139 @@ If you do not want to develop your plugin using Golang, you can follow its stand
...
 Adopting these standards can bring significant benefits, such as joining forces on maintaining the common standards as the features provided by Kubebuilder and take advantage of the contributions made by the community. This allows you to focus on the specific needs and requirements for your plugin and use-case.
 
 And then, you will also be able to use custom plugins and options currently or in the future which might to be provided by these projects as any other which decides to persuade the same standards.
 
 </aside>
 
-## Custom Plugins 
+## Custom Plugins
 
 Note that users are also able to use plugins to customize their scaffold and address specific needs. See that Kubebuilder provides the [declarative][declarative-code] plugin which can be used when for example an API is scaffold:
```
[the remark this answers is not in the document]
Comment (Kavinjsir) [reply docs/book/src/plugins/creating-plugins.md]: Yeah, sounds good!

File: docs/book/src/plugins/creating-plugins.md [hunk 6]
```diff
@@ -25,54 +25,93 @@ If you do not want to develop your plugin using Golang, you can follow its stand
```
[the remark this answers is not in the document]
Comment (Kavinjsir) [reply docs/book/src/plugins/creating-plugins.md:25]: +1
What about another section of testing your plugin?

**Review by camilamacedo86 — APPROVED** [review]
It seems great for me 🥇 
We can improve in follow-up always.

So, let's move with this one. 

/lgtm

**Pull request merged**